src/Microsoft.VisualStudio.Composition/Configuration/SerializationContextBase.cs

```diff
@@ -677,16 +673,62 @@ protected ImmutableArray<T> ReadImmutableArray<T>(BinaryReader reader, Func<T> i
 
                 if (count == 0)
                 {
-                    return ImmutableArray<T>.Empty;
+                    return Array.Empty<T>();
+                }
+
+                var list = new T[count];
+                for (int i = 0; i < list.Length; i++)
+                {
+                    list[i] = itemReader();
+                }
+
+                return list;
+            }
+        }
+
+        protected ImmutableArray<TypeRef?> ReadTypeRefImmutableArray(BinaryReader reader, Func<TypeRef?> itemReader)
+        {
+            using (this.Trace(typeof(TypeRef).Name, isArray: true))
+            {
+                uint count = this.ReadCompressedUInt();
+
+                switch (count)
+                {
+                    case 0:
+                        return ImmutableArray<TypeRef?>.Empty;
+                    case 1:
+                        return ImmutableArray.Create(itemReader());
+                    case 2:
+                        return ImmutableArray.Create(itemReader(), itemReader());
+                    case 3:
+                        return ImmutableArray.Create(itemReader(), itemReader(), itemReader());
+                    case 4:
+                        return ImmutableArray.Create(itemReader(), itemReader(), itemReader(), itemReader());
+                }
+
+                if (count > 0xffff)
+                {
+                    // Probably either file corruption or a bug in serialization.
+                    // Let's not take untold amounts of memory by throwing out suspiciously large lengths.
+                    throw new NotSupportedException();
```

Comment:
nit: NotSupportedException sounds a bit misleading. An InvalidOperationException with a message that contains the suspicious `count` value would be far more useful if this ever showed up in telemetry or logs.

Reply:
This was a direct copy from ReadList/ReadArray

Reply:
Fair enough. :)